neutrinordp/xrdp-neutrinordp.c

```diff
@@ -609,6 +610,77 @@ lxrdp_set_param(struct mod *mod, const char *name, const char *value)
         {
             mod->perf_settings_values_mask |= PERF_DISABLE_CURSOR_SHADOW;
         }
+    }
+        else if (g_strcmp(name, "neutrinordp.allow_client_keyboardLayout") == 0)
+    {
+        if (g_text2bool(value))
+        {
+            if (settings->kbd_type == 0)
+            {
+                settings->kbd_type = mod->client_info.keyboard_type;
+            }
+            if (settings->kbd_subtype == 0)
+            {
+                settings->kbd_subtype = mod->client_info.keyboard_subtype;
+            }
+            if (settings->kbd_fn_keys == 0)
+            {
+                /* Define the most common number of function keys, 12. 
+                   because we can't get it from client. */
+                settings->kbd_fn_keys = 12;
+            }
+            if (settings->kbd_layout == 0)
+            {
+                settings->kbd_layout = mod->client_info.keylayout;
```

Comment:
I don't understand the need for this mask, so I've had a look around.

I'm still a bit confused, but it seems that some keyboards may need this mask and some may not.

The keyboard layouts are currently [here](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/windows-language-pack-default-values) on the MS website. Or at least some of them are. There's only one Japanese one listed, and it's `0x00000411`. The US ones however are:-

Layout | Identifier
-- | --
United States - English | 0x00000409
United States - Dvorak | 0x00010409
United States - International | 0x00020409
United States-Dvorak for left hand | 0x00030409
United States-Dvorak for right hand | 0x00040409

So now I understand your question about US Dvorak keyboards, although I'm not qualified to answer it.

Is it the case that Japanese keyboards will always need this mask?

Reply:
Thank you for all your research.

Yes, it's true.
That's the complication with Japanese keyboards.

@metalefty  is the leading expert on this issue in xrdp.
Thanks to his efforts over the years, xrdp is now more than usable in a Japanese environment.

@metalefty has published his work in Japanese.
https://xrdp.vmeta.jp/%E6%B4%BB%E5%8B%95%E5%AE%9F%E7%B8%BE?action=AttachFile&do=view&target=osc2016-tokyo-spring-xrdp-keylayout. pdf

I have referred to this.

Reply:
@metalefty  Could you please add more information?
Also, could you please point out if I'm wrong?

Reply:
Regarding the US keylayout value, the lower 16 bits imply the language and the upper 16 bits imply layout variants like QWERTY or Dvorak, etc. 

The Japanese keylayout is a little bit different. Japanese keyboard has a single layout but we Japanese use several kinds of IMEs. Ordinary people only use Microsoft's one but some people install and use the third-party one. I see different upper 16 bits are sent from client to server depending on the active IME.
But regardless of the upper 16 bits value, keyboard layouts are same. 

Layout | Identifier
-- | --
Japanese - Direct input | 0x00000411
Japanese - Microsoft IME | 0x00010411
Japanese - Third-party IME | 0xE02n0411

`n` will be incremented by installed order of third-party IMEs.

The mask is a hack to treat as same with the representative keylayout `0x00000411` if any different upper 16 bits are sent from the client.

> I'm still a bit confused, but it seems that some keyboards may need this mask and some may not.

I'm also confused but actually, some keyboards may need this mask. I couldn't find any official MS documents about the upper 16 bits of keylayout value. I'm still looking for that.

Reply:
Thanks for that - it helps a bit.

In that case, I can't see any reason for now why we shouldn't hard-code the mask for a keyboard type of 7. There will be an override anyway, so should we come across another type which needs a mask (probably 0xffff again) we can add it after client testing.

Reply:
Thank you, @matt335672  @metalefty !

@metalefty ,Thank you for supplementing my poor explanation.

As @matt335672  said, I thought hardcoding `keyboardLayout` value as `0x00000411` for Japanese keyboard (`kbd_type:0x07`) would make my intention clearer.
Once I've done that, I'm coding with that policy.